ocsp.go

```diff
@@ -584,44 +607,62 @@ func validateWithCacheForAllCertificates(verifiedChains []*x509.Certificate) boo
 		subject := verifiedChains[j]
 		issuer := verifiedChains[j+1]
 		status, _, _ := validateWithCache(subject, issuer)
-		if !isValidOCSPStatus(status) {
+		if !isValidOCSPStatus(status.code) {
 			return false
 		}
 	}
 	return true
 }
 
-func validateWithCache(subject, issuer *x509.Certificate) (ocspStatusCode, []byte, *certIDKey) {
+func validateWithCache(subject, issuer *x509.Certificate) (*ocspStatus, []byte, *certIDKey) {
 	ocspReq, err := ocsp.CreateRequest(subject, issuer, &ocsp.RequestOptions{})
 	if err != nil {
 		glog.V(2).Infof("failed to create OCSP request from the certificates.\n")
-		return ocspFailedComposeRequest, nil, nil
+		return &ocspStatus{
+			code: ocspFailedComposeRequest,
+			err:  errors.New("failed to create a OCSP request"),
+		}, nil, nil
 	}
 	encodedCertID, ocspS := extractCertIDKeyFromRequest(ocspReq)
 	if ocspS.code != ocspSuccess {
 		glog.V(2).Infof("failed to extract CertID from OCSP Request.\n")
-		return ocspFailedComposeRequest, ocspReq, nil
+		return &ocspStatus{
+			code: ocspFailedComposeRequest,
+			err:  errors.New("failed to extract cert ID Key"),
+		}, ocspReq, nil
 	}
-	ocspValidatedWithCache := checkOCSPResponseCache(encodedCertID, subject, issuer)
-	return ocspValidatedWithCache.code, ocspReq, encodedCertID
+	status := checkOCSPResponseCache(encodedCertID, subject, issuer)
+	return status, ocspReq, encodedCertID
 }
 
 func downloadOCSPCacheServer() {
 	if strings.EqualFold(os.Getenv(cacheServerEnabledEnv), "false") {
 		glog.V(2).Infof("skipping downloading OCSP Cache.")
 		return
 	}
-	ocspClient := &http.Client{
-		Timeout:   30 * time.Second,
-		Transport: snowflakeInsecureTransport,
+	ocspCacheServerURL := os.Getenv(cacheServerURLEnv)
+	if ocspCacheServerURL == "" {
+		ocspCacheServerURL = fmt.Sprintf("%v/%v", cacheServerURL, cacheFileBaseName)
 	}
-	ocspURL := fmt.Sprintf("%v/%v", cacheServerURL, cacheFileBaseName)
-	u, err := url.Parse(ocspURL)
+	u, err := url.Parse(ocspCacheServerURL)
 	if err != nil {
 		return
 	}
-	glog.V(2).Infof("downloading OCSP Cache from server %v", ocspURL)
-	ret, ocspStatus := retryOCSPCacheServer(ocspClient, http.NewRequest, u, retryOCSPTimeout, retryOCSPHTTPTimeout)
+	glog.V(2).Infof("downloading OCSP Cache from server %v", ocspCacheServerURL)
+	timeoutStr := os.Getenv(ocspTestResponseCacheServerTimeoutEnv)
+	timeout := defaultOCSPCacheServerTimeout
+	if timeoutStr != "" {
+		var timeoutInMilliseconds int
+		timeoutInMilliseconds, err = strconv.Atoi(timeoutStr)
+		if err == nil {
+			timeout = time.Duration(timeoutInMilliseconds) * time.Millisecond
+		}
+	}
+	ocspClient := &http.Client{
+		Timeout:   timeout,
```

Comment:
Shouldn't this be a smaller timeout than OCSPCacheServerTimeout to allow for multiple retry to OCSP cache server?